### ISSUE TITLE
feature: add `router` option to dynamically route to different `target`

### ISSUE THREAD

**Issue by chimurai** [chimurai/http-proxy-middleware#54]
Allow user to provide a function to (re)route to a different target; context information from the `request` object can be used to determine the target as user wishes.

```javascript
var apiProxy = proxy('/api', {
  target: 'http://localhost:1234',
  router: function (req) {
    return 'http://newtarget.localhost:5000';
    // return;  // to use default `target`
  }
});
```

Deprecate `proxyTable` in favor of `router`. The `proxyTable` functionality can still be used by providing a map (proxyTable) with routing rules.

```javascript
var proxyTable = {
    "integration.localhost:3000" : "http://localhost:8001",    // host only
    "staging.localhost:3000"     : "http://localhost:8002",    // host only
    "localhost:3000/api"         : "http://localhost:8003",    // host + path
    "/rest"                      : "http://localhost:8004"     // path only
};

var apiProxy = proxy('/api', {
  target: 'http://localhost:1234',
  router: proxyTable
});
```

References:
- https://blog.nodejitsu.com/node-http-proxy-1dot0/
- https://github.com/nodejitsu/node-http-proxy/issues/571
- https://github.com/nodejitsu/node-http-proxy/issues/827

**Comment by chimurai** [comment]
Leave a reply if you can improve this proposal to cover a certain use-case.

**Comment by chimurai** [comment]
published in [v0.16.0](https://github.com/chimurai/http-proxy-middleware/releases/tag/v0.16.0)
